substrate/frame/assets-holder/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+	pallet_prelude::*,
+	traits::{tokens::IdAmount, VariantCount, VariantCountOf},
+	BoundedVec,
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod impls;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config(with_default)]
+	pub trait Config<I: 'static = ()>:
+		frame_system::Config + pallet_assets::Config<I, Holder = Pallet<Self, I>>
+	{
+		/// The overarching freeze reason.
+		#[pallet::no_default_bounds]
+		type RuntimeHoldReason: Parameter + Member + MaxEncodedLen + Copy + VariantCount;
+
+		/// The overarching event type.
+		#[pallet::no_default_bounds]
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// Number of holds on an account would exceed the count of `RuntimeHoldReason`.
+		TooManyHolds,
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		// `who`s held balance was increased by `amount`.
+		Held {
+			who: T::AccountId,
+			asset_id: T::AssetId,
+			reason: T::RuntimeHoldReason,
+			amount: T::Balance,
+		},
+		// `who`s held balance was decreased by `amount`.
+		Released {
+			who: T::AccountId,
+			asset_id: T::AssetId,
+			reason: T::RuntimeHoldReason,
+			amount: T::Balance,
+		},
+		// `who`s held balance was burned by `amount`.
+		Slashed {
+			who: T::AccountId,
+			asset_id: T::AssetId,
+			reason: T::RuntimeHoldReason,
+			amount: T::Balance,
+		},
+	}
+
+	/// A map that stores holds applied on an account for a given AssetId.
+	#[pallet::storage]
+	pub(super) type Holds<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AssetId,
+		Blake2_128Concat,
+		T::AccountId,
+		BoundedVec<
+			IdAmount<T::RuntimeHoldReason, T::Balance>,
+			VariantCountOf<T::RuntimeHoldReason>,
+		>,
+		ValueQuery,
+	>;
+
+	/// A map that stores the current total held balance for every account on a given AssetId.
+	#[pallet::storage]
+	pub(super) type HeldBalances<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AssetId,
+		Blake2_128Concat,
+		T::AccountId,
+		T::Balance,
+	>;
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+		#[cfg(feature = "try-runtime")]
+		fn try_state(_: BlockNumberFor<T>) -> Result<(), sp_runtime::TryRuntimeError> {
+			Self::do_try_state()
+		}
+	}
+}
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	#[cfg(any(test, feature = "try-runtime"))]
+	fn do_try_state() -> Result<(), sp_runtime::TryRuntimeError> {
+		use sp_runtime::traits::{Saturating, Zero};
+
+		for (asset, who, _) in HeldBalances::<T, I>::iter() {
+			let held_amount: T::Balance = Holds::<T, I>::get(asset.clone(), who.clone())
+				.iter()
+				.map(|l| l.amount)
```

Comment:
we could ensure that this amount is not zero.

Reply:
This is ensured [here](https://github.com/pandres95/polkadot-sdk/blob/e7155eb5279b26606a0f95166252bffabed43b50/substrate/frame/assets-holder/src/impls.rs#L182)

Reply:
Yes the code is correct, but we could test it is correct here as well. anyway it is fine to me

Reply:
Ah! I see. This `try_state` is being tested [here](https://github.com/pandres95/polkadot-sdk/blob/818e59048049ce51c6c3dca860653e533582769a/substrate/frame/assets-holder/src/mock.rs#L152)

Reply:
I wanted to say to write something like this:

(it doesn't compile, but it was the idea)
```rust
.map(|l| {ensure!(l.amount != Zero::zero(), "zero amount is invalid"); l.amount})
```